### PR TITLE
NFT for Royale sign up

### PR DESCRIPTION
**Pull request by grujawork**
Some open questions for elaborate:

- added first draft version of NFT who needs a `tokenURI` to be set, so we need URI which points to JSON file (probably)

- do we mint them with same URI until we change price or something ?! or we mint them with second input parameter in mint method ?!

- elaborate about price, now price is set in a NFT contract `pricePaidForVoucher` and that price is send to Royale Contract when user is signed, burning 'one to one' relation with no checking any of price just send sUSD for which voucher is payed in that time, what if price of voucher is paid more or less then buy-in (in a case that voucher is less then a buy in do we burn that NFT + ask for the difference in a sUSD and in a case of voucher is more then buy in amount do we just send sUSD and decrease `pricePaidForVoucher` and NFT stays ?! )

- automate buy-in with price in voucher do we do it manually or we do some 'whitelist addresses' in NFT, example in Royale when setting buy-in call setPrice in NFT but only from that address or owner address

### CODE REVIEW

File: contracts/ThalesRoyale/ThalesRoyale.sol [hunk 1]
```diff
@@ -82,26 +84,27 @@ contract ThalesRoyale is Initializable, ProxyOwned, PausableUpgradeable, ProxyRe
...
-    ) public initializer {
+        uint _pauseBetweenSeasonsTime    
+        ) 
+        public initializer {
         setOwner(_owner);
         initNonReentrant();
         oracleKey = _oracleKey;
         priceFeed = _priceFeed;
         rewardToken = IERC20Upgradeable(_rewardToken);
+        royaleVoucher = IThalesRoyaleVoucher(_royaleVoucher);
```
Comment (danijelthales) [inline contracts/ThalesRoyale/ThalesRoyale.sol:101]: afaik initializer method cant change as the proxy is already deployed.

Comment (danijelthales) [reply contracts/ThalesRoyale/ThalesRoyale.sol:101]: roayeVoucher should be set through a setter

File: contracts/ThalesRoyale/ThalesRoyaleVoucher.sol [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        _setTokenURI(newItemId, tokenURI);
+
+        return newItemId;
+    }
+
+    function burn(uint tokenId) external canBeBurned(tokenId) {
+        super._burn(tokenId);
+    }
+
+    function burnWithTransfer(uint tokenId) external canBeBurned(tokenId) {
```
Comment (danijelthales) [inline contracts/ThalesRoyale/ThalesRoyaleVoucher.sol:70]: this is sent from Royale contract, how does it pass the isApprovedOrOwner check?

File: contracts/ThalesRoyale/ThalesRoyaleVoucher.sol [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    string public _name = "Thales Royale Voucher";
+    string public _symbol = "TRV";
+    bool public paused = false;
+    string public tokenURI;
+
+    address public thalesRoyaleAddress;
+
+    IERC20 public sUSD;
+    uint public price;
+    mapping(uint => uint) public pricePerVoucher;
```
Comment (danijelthales) [inline contracts/ThalesRoyale/ThalesRoyaleVoucher.sol:30]: NFTs are supposed to be stateful, cant this be a part of the minted NFT storage?

File: contracts/ThalesRoyale/ThalesRoyaleVoucher.sol [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    using SafeERC20 for IERC20;
+
+    /* ========== STATE VARIABLES ========== */
+
+    Counters.Counter private _tokenIds;
+
+    string public _name = "Thales Royale Voucher";
+    string public _symbol = "TRV";
+    bool public paused = false;
+    string public tokenURI;
```
Comment (danijelthales) [inline contracts/ThalesRoyale/ThalesRoyaleVoucher.sol:24]: I dont know what tokenURI is for to answer the question whether it should be the same

File: contracts/ThalesRoyale/ThalesRoyale.sol (hunk 1 above)
Comment (grujawork) [reply contracts/ThalesRoyale/ThalesRoyale.sol:101]: ah ok, will do

File: contracts/ThalesRoyale/ThalesRoyaleVoucher.sol (hunk 2 above)
Comment (grujawork) [reply contracts/ThalesRoyale/ThalesRoyaleVoucher.sol:70]: there is a `function approve(address to, uint256 tokenId)` in ERC721 contract, so user will need to approve that royale contract "use" his voucher

**Comment by danijelthales** [comment]
> Some open questions for elaborate:
> 
> * added first draft version of NFT who needs a `tokenURI` to be set, so we need URI which points to JSON file (probably)
> * do we mint them with same URI until we change price or something ?! or we mint them with second input parameter in mint method ?!
> * elaborate about price, now price is set in a NFT contract `pricePaidForVoucher` and that price is send to Royale Contract when user is signed, burning 'one to one' relation with no checking any of price just send sUSD for which voucher is payed in that time, what if price of voucher is paid more or less then buy-in (in a case that voucher is less then a buy in do we burn that NFT + ask for the difference in a sUSD and in a case of voucher is more then buy in amount do we just send sUSD and decrease `pricePaidForVoucher` and NFT stays ?! )
> * automate buy-in with price in voucher do we do it manually or we do some 'whitelist addresses' in NFT, example in Royale when setting buy-in call setPrice in NFT but only from that address or owner address

Yeah use the same tokenURI which you will host on IPFS.
Ignore any complexity around the price for now, make the buy-in amount changeable by owner in NFT contract. We will transfer more sUSD for any remaining vouchers that have not been burned if the buy-in increases

File: contracts/ThalesRoyale/ThalesRoyaleVoucher.sol (hunk 2 above)
Comment (danijelthales) [reply contracts/ThalesRoyale/ThalesRoyaleVoucher.sol:70]: not really that elegant, you should be able to just check that the caller is the Royale address and allow the burn

**Comment by grujawork** [comment]
> > Some open questions for elaborate:
> > 
> > * added first draft version of NFT who needs a `tokenURI` to be set, so we need URI which points to JSON file (probably)
> > * do we mint them with same URI until we change price or something ?! or we mint them with second input parameter in mint method ?!
> > * elaborate about price, now price is set in a NFT contract `pricePaidForVoucher` and that price is send to Royale Contract when user is signed, burning 'one to one' relation with no checking any of price just send sUSD for which voucher is payed in that time, what if price of voucher is paid more or less then buy-in (in a case that voucher is less then a buy in do we burn that NFT + ask for the difference in a sUSD and in a case of voucher is more then buy in amount do we just send sUSD and decrease `pricePaidForVoucher` and NFT stays ?! )
> > * automate buy-in with price in voucher do we do it manually or we do some 'whitelist addresses' in NFT, example in Royale when setting buy-in call setPrice in NFT but only from that address or owner address
> 
> Yeah use the same tokenURI which you will host on IPFS. Ignore any complexity around the price for now, make the buy-in amount changeable by owner in NFT contract. We will transfer more sUSD for any remaining vouchers that have not been burned if the buy-in increases

Cool

**Review by danijelthales — CHANGES_REQUESTED** [review]

File: scripts/deployThalesRoyale/thalesRoyaleVoucher/deployVoucher.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    const uri = 'http://my-json-server.typicode.com/abcoathup/samplenft/tokens/0';
+
+    const ThalesRoyaleVoucher = await ethers.getContractFactory('ThalesRoyaleVoucher');
+    const ThalesRoyaleVoucherDeployed = await ThalesRoyaleVoucher.deploy(
+        sUSDAddress,
+        price,
+        uri
+    );
+    await ThalesRoyaleVoucherDeployed.deployed();
+    setTargetAddress('ThalesRoyaleVoucher', network, ThalesRoyaleVoucherDeployed.address);
```
[the remark this answers is not in the document]
Comment (grujawork) [reply scripts/deployThalesRoyale/thalesRoyaleVoucher/deployVoucher.js]: done

File: scripts/deployThalesRoyale/thalesRoyaleVoucher/deployVoucher.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    console.log('Account is: ' + owner.address);
+    console.log('Network:' + network);
+
+    sUSDAddress = getTargetAddress('ProxysUSD', network);
+    console.log('ProxysUSD :', sUSDAddress);
+
+    const price = w3utils.toWei('30');
+    const uri = 'http://my-json-server.typicode.com/abcoathup/samplenft/tokens/0';
```
[the remark this answers is not in the document]
Comment (grujawork) [reply scripts/deployThalesRoyale/thalesRoyaleVoucher/deployVoucher.js]: will do

**Review by danijelthales — APPROVED** [review]

**Pull request merged**